File: apps/mobile/src/platform/biometrics.ts
/**
 * `BiometricProvider` — Android Keystore + BiometricPrompt.
 *
 * Как это устроено (подробности — в `android/.../Biometrics.kt`):
 *
 *   enroll  — в Keystore заводится AES-ключ с `setUserAuthenticationRequired`,
 *             им шифруется секрет, шифротекст ложится в приватный каталог
 *             приложения. Сам секрет Keystore не покидает открытым.
 *   unlock  — `BiometricPrompt` с `CryptoObject`; ключ становится пригоден к
 *             использованию только после успешной биометрии. Отмена
 *             пользователем — это `null`, а не ошибка (BEHAVIOR §5.2:
 *             «Отмена биометрии пользователем — не ошибка»).
 *
 * **Честность важнее галочки.** `isAvailable()` возвращает `true` только когда
 * на устройстве есть настроенная стойкая биометрия и API её поддерживает
 * (Android 9+). На всём остальном — `false`, и UI **скроет** тумблер
 * (BEHAVIOR §5.1). Ни одной имитации: провайдер, который «как бы» защищает
 * ключ, хуже отсутствующего — пользователь считает заметки защищёнными.
 */
import type { BiometricProvider } from '@zapiski/core';

import { COMMANDS, call } from './ipc';

/**
 * Провайдер — или `null`, если на устройстве биометрии нет.
 *
 * Проверка делается ЗДЕСЬ, до сборки возможностей, и это починка сообщённого
 * отказа. Раньше порт заявлялся всегда, а `PlatformCapabilities.biometrics` —
 * поле, по которому интерфейс решает, показывать ли тумблер. Лист шифрования
 * смотрел ровно на него и показывал «Использовать отпечаток» на любом
 * устройстве — даже там, где стойкой биометрии нет вовсе. Дальше человек жал
 * «Зашифровать» и получал крах.
 *
 * Windows делает так с самого начала (`apps/desktop/src/platform/biometrics.ts`):
 * без настроенного Hello порт — `null`. Расхождение между оболочками и было
 * дырой: каждый потребитель порта обязан был помнить про `isAvailable()`, и
 * один из двух не помнил.
 */
export async function createBiometrics(): Promise<BiometricProvider | null> {
  const provider = createProvider();
  return (await provider.isAvailable()) ? provider : null;
}

function createProvider(): BiometricProvider {
  return {
    async isAvailable() {
      try {
        return await call<boolean>(COMMANDS.biometricsAvailable);
      } catch {
        return false;
      }
    },

    async enroll(keyId, secret) {
      // Ошибку пробрасываем: `packages/app` включает тумблер только после
      // успешного enroll, иначе пользователь считал бы, что вход по отпечатку
      // настроен, а он бы не работал.
      await call<void>(COMMANDS.biometricsEnroll, {
        keyId,
        secret: Array.from(secret),
      });
    },

    async unlock(keyId) {
      const bytes = await call<number[] | null>(COMMANDS.biometricsUnlock, { keyId });
      return bytes === null ? null : Uint8Array.from(bytes);
    },

    async remove(keyId) {
      await call<void>(COMMANDS.biometricsRemove, { keyId });
    },
  };
}

File: apps/mobile/src-tauri/src/biometrics.rs
//! Реализация порта `BiometricProvider`.
//!
//! Вся работа с Keystore и `BiometricPrompt` — в Kotlin
//! (`android/.../Biometrics.kt`), здесь только команды и приведение типов.
//!
//! Важное правило (BEHAVIOR §5.1—5.2), которое видно уже отсюда:
//!
//! * `biometrics_available` возвращает `false`, если стойкой биометрии на
//!   устройстве нет или её не поддерживает версия Android. Тогда UI **скроет**
//!   тумблер. Возвращать `true` «на всякий случай» нельзя: пользователь решит,
//!   что заметка защищена отпечатком, а она не будет;
//! * отмена биометрии пользователем — это `Ok(None)`, а не ошибка. UI покажет
//!   поле пароля без единого сообщения.

/// Есть ли на устройстве настроенная стойкая биометрия.
#[tauri::command(async)]
pub fn biometrics_available() -> bool {
    // Ошибка моста — это тоже «нет биометрии». Показывать тумблер, который
    // упадёт при нажатии, хуже, чем не показывать его вовсе.
    crate::android::biometrics_available().unwrap_or(false)
}

/// Положить секрет под биометрию.
#[tauri::command(async)]
pub fn biometrics_enroll(key_id: String, secret: Vec<u8>) -> Result<(), String> {
    if secret.is_empty() {
        return Err("пустой секрет".to_owned());
    }
    crate::android::biometrics_enroll(&key_id, &secret)
}

/// Достать секрет после успешной биометрии. `None` — пользователь отменил.
#[tauri::command(async)]
pub fn biometrics_unlock(key_id: String) -> Result<Option<Vec<u8>>, String> {
    crate::android::biometrics_unlock(&key_id)
}

/// Убрать ключ: снятие шифрования или выключение тумблера.
#[tauri::command(async)]
pub fn biometrics_remove(key_id: String) -> Result<(), String> {
    crate::android::biometrics_remove(&key_id)
}
